tests/unit/gh_pr_upsert/git_test.py

```diff
@@ -0,0 +1,255 @@
+
+        pr = git.pr()
+
+        run.assert_called_once_with(
+            ["gh", "pr", "view", "--json", "state,commits"], json=True
+        )
+        assert pr == run.return_value
+
+    def test_it_returns_None_if_theres_no_open_pr(self, run):
+        run.return_value = {"state": "CLOSED"}
+
+        assert git.pr() is None
+
+    def test_it_returns_None_if_the_command_fails(self, run):
+        run.side_effect = CalledProcessError(
+            returncode=sentinel.returncode, cmd=sentinel.cmd
+        )
+
+        assert git.pr() is None
+
+    @pytest.fixture(autouse=True)
+    def cache_clear(self):
+        git.pr.cache_clear()
+
+
+class TestGitHubRepo:
+    def test_it(self, run):
+        repo = git.github_repo()
+
+        run.assert_called_once_with(
+            ["gh", "repo", "view", "--json", "url,sshUrl,defaultBranchRef"], json=True
+        )
+        assert repo == run.return_value
+
+    @pytest.fixture(autouse=True)
+    def cache_clear(self):
+        git.github_repo.cache_clear()
+
+
+class TestDefaultBranch:
+    def test_it(self, run):
+        run.return_value = {"defaultBranchRef": {"name": sentinel.default_branch_name}}
+
+        branch_name = git.default_branch()
+
+        run.assert_called_once_with(
+            ["gh", "repo", "view", "--json", "url,sshUrl,defaultBranchRef"], json=True
+        )
+        assert branch_name == sentinel.default_branch_name
+
+    @pytest.fixture(autouse=True)
+    def cache_clear(self):
+        git.github_repo.cache_clear()
+
+
+class TestPushURLs:
+    def test_it(self, run):
+        run.return_value = {"url": "test_url", "sshUrl": "test_ssh_url"}
+
+        urls = git.push_urls()
+
+        run.assert_called_once_with(
+            ["gh", "repo", "view", "--json", "url,sshUrl,defaultBranchRef"], json=True
+        )
+        assert urls == ["test_url.git", "test_ssh_url"]
+
+    @pytest.fixture(autouse=True)
+    def cache_clear(self):
+        git.github_repo.cache_clear()
+
+
+class TestCurrentBranch:
+    def test_it(self, run):
+        current_branch = git.current_branch()
+
+        run.assert_called_once_with(
+            ["git", "symbolic-ref", "--quiet", "--short", "HEAD"]
+        )
+        assert current_branch == run.return_value
+
+    @pytest.fixture(autouse=True)
+    def cache_clear(self):
+        git.current_branch.cache_clear()
+
+
+class TestCommittedChanges:
+    def test_it(self, run):
+        diff = git.committed_changes()
+
+        run.assert_called_once_with(["git", "diff", "origin/main..."])
+        assert diff == run.return_value
+
+    @pytest.fixture(autouse=True)
+    def remote(self, mocker):
+        return mocker.patch(
+            "gh_pr_upsert.git.remote", autospec=True, return_value="origin"
+        )
+
+    @pytest.fixture(autouse=True)
+    def default_branch(self, mocker):
+        return mocker.patch(
+            "gh_pr_upsert.git.default_branch", autospec=True, return_value="main"
+        )
```

Comment:
In this and some other tests below I decided to mock out the other functions from the same file that the function-under-test calls. It gets really horrible if you don't do this because you have to set up lots of complicated return values for Git and GitHub CLI commands that are run by functions that are only indirectly called by the function under test. The more isolated approach is much better in this case.